Layout

josemigallas opened the pull request:
App implementation has been integrated into layout, allowing to generate a tournament by means of some forms, button and a table.

josemigallas committed:
creates basic layout for players enrollment

josemigallas committed:
adds jquery dependency

josemigallas committed:
adds jasmine-es6 dependency

josemigallas committed:
adds class Player with tests

josemigallas committed:
adds class Tournament with some test

josemigallas committed:
adds class Match with tests

josemigallas committed:
class Tournament generates a set of rounds with matches based on the amount players

josemigallas committed:
refactors class Tournament and tests in a neater way

josemigallas committed:
adds file utils.js for wrapping common functions up

josemigallas committed:
implements ramdom player pairing for 1st round

josemigallas committed:
fixes uncompleted test

josemigallas committed:
minor improovements in layout responsiveness

josemigallas committed:
moves src folder content to project root directory

josemigallas committed:
changes player list layout from a simple list to a neater table

josemigallas committed:
Merge branch 'app' into layout

josemigallas committed:
allows players to enroll and updates table layout

josemigallas committed:
fixes error with test: generating a tournament should assign player objects to matches, not arrays.

josemigallas committed:
Merge branch 'app' into layout

josemigallas committed:
allows tournament to be generated, closing enrollment

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+        mTournament.addPlayer(newPlayer);
+        updateTable(newPlayer);
+
+    } catch (e) {
```
Don´t use exceptions to drive your logic. By reading this, I don´t know who is expected to throw. Use return instead.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+    $("#buttonCreate").click(generateTournament);
+});
+
+function addPlayerIfValid() {
```
This is more than I would expect in app.js, it should just initialize things and kick off the app.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+    $("#playerListBody").append(newRow(player));
+}
+
+var newRow = (player) => `<tr><td>${player.name.replace(/</g, "&lt;").replace(/>/g, "&gt;")}</td><td>${player.rank}</td></tr>`;
```
Arrow syntax is not adding value here IMO.

pablonete replied:
And newRow is not a good name for a function. Use a verb like composeHtmlForPlayerRow().

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+}
+
+function addPlayerToTable(player) {
+    $("#playerListBody").append(newRow(player));
```
There´s a lot of responsibilities on this module: 1. App start up, 2. Add player, 3. Show players

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+import $ from "jquery";
+import * as tournament from "./tournament/tournament";
+import * as player from "./tournament/player";
```
On these 2 cases, I think it´s more natural to import {Player} and {Tournament} directly.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+import * as tournament from "./tournament/tournament";
+import * as player from "./tournament/player";
+
+var mTournament = new tournament.Tournament();
```
Avoid the m prefix.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+    }
+}
+
+function updateTable(player) {
```
This method is distracting and has a weak boundary, I´d inline it.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+function getPlayerInfo() {
+    return {
+        name: $("#fieldName")[0].value,
+        rank: $("#fieldRank")[0].value
```
Typically in jQuery you don´t need to access [0]. By using .val() you can get or set the value of the elements, which will work for 1 or more elements.

pablonete commented on app/app.js:
```diff
@@ -0,0 +1,61 @@
+        closeEnrollment();
+
+    } catch (e) {
+        window.alert("Too few players!");
```
This is one of the reasons we should not use exceptions. Many other things can go wrong, and you´ll end up showing the wrong message.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+        var mTournament = new tournament.Tournament();
+        mTournament.addPlayer(new player.Player("Spiderman", 1));
+        mTournament.addPlayer(new player.Player("Batman", 1));
+        mTournament.addPlayer(new player.Player("Ironman", 1));
```
Is it valid to have 3 players with the same rank? It´s not the best way to specify your algorithm.

josemigallas replied:
Do have name and rank be both unique for each player?

pablonete replied:
I think so. I´ve never seen 2 players with the same rank.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+        expect(mTournament.hasProperNumberOfPlayers()).toEqual(false);
+    });
+
+    it('a total of 4 players should not be a proper number', function() {
```
This title is wrong.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+        mTournament.addPlayer(new player.Player("Batman", 1));
+        mTournament.addPlayer(new player.Player("Ironman", 1));
+        mTournament.generate();
+        expect(mTournament.players.length).toEqual(4);
```
Then why hasProperNumberOfPlayers() was returning false for 3 players above? It seems contradictory.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        return this.players.length - 1;
+    }
+
+    generateTournamentBracket() {
```
I suggest breaking up this module into the tournament container and the bracket generator.

pablonete commented on app/tournament/player.js:
```diff
@@ -0,0 +1,27 @@
+    }
+}
+
+export class PlayerError extends Error {
```
What do you need this class for? I think validation errors don´t need to be modeled as exceptions.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,20 @@
+    }
+}
+
+export class MatchException extends Match {
```
This is very wrong, not just that I don´t feel we need an error class, it´s that it´s extending from Match.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,20 @@
+        if (this.players.length >= 2) {
+            throw new MatchException("Match has already 2 players");
+        }
+        this.players.push(player);
```
I´m curious why players are managed through this method while parents not.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,20 @@
+export class Match {
+    constructor() {
+        this.round = 1;
```
It´s odd to set a value on the constructor to be overwritten immediately.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+                return true;
+            }
+        };
+        return false;
```
Use Array.find()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+
+    addPlayer(player) {
+        if (this.existsAlready(player)) {
+            throw new TournamentError(`Duplicate player: "${player.name}".`);
```
Avoid exceptions. This is correct (I would just use Error), but it must be avoided by checking exists from outside.

josemigallas replied:
So after adding some code validation outside, should I clear my exceptions off my modules or just the try/catch blocks?

pablonete replied:
No, you can keep the throw, it´s just that your logic must avoid it. If it happens, you have a bug, so you don´t even need to catch it, you have to fix the bug.

It´s like a/b, I don´t want to catch the DivisionByZero error, I just want to ensure that everytime I make to that line, b is not zero. If it is, I have a bug.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        this.players.push(player);
+    }
+
+    existsAlready(player) {
```
Rename to existsPlayer()

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        this.pairUpPlayers();
+    }
+
+    checkNumberOfPlayers() {
```
Try to avoid publishing private methods. Move up outside the class, and pass this as an argument.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+
+    enaughPlayers() {
```
Typo, but also bad name. AreEnoughPlayers?

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        return this.players.length >= 2;
+    }
+
+    hasProperNumberOfPlayers() {
```
Do you want to unit test this logic? Move to a public utility method. Tournament doesn´t need to publish it.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    fillUpWithGhosts() {
+        var goodNumber = parseInt(Math.pow(10, this.players.length.toString(2).length), 2);
```
Extract a well-named method for this logic, and try to make it more understandable.  
Using 10^n to compose a small number is overkill, is not simpler a for(i*=2) until you overpass the number?

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        var goodNumber = parseInt(Math.pow(10, this.players.length.toString(2).length), 2);
+        var i = 1;
+        while (this.players.length < goodNumber) {
+            this.addPlayer(new player.Player(`Ghost${i++}`, 1));
```
Ghost players should have minimum ranks, and perhaps without dupes.

pablonete replied:
Ghost players can be used by the generate algorithm, but they should never get added to the tournament players.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+
+    fillUpWithGhosts() {
+        var goodNumber = parseInt(Math.pow(10, this.players.length.toString(2).length), 2);
+        var i = 1;
```
i = 1 plus while() plus i++ makes up a for loop, for the sake of readability.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        var totalRounds = this.totalRounds();
+
+        this.matches.forEach( function(match, i, matches) {
+            var round = totalRounds - parseInt(Math.log2(i + 1));
```
The indexes make up a complex logic, based on log, +1 increments... This is very error prone.
Have you considered creating matches from previous matches like:
new Match([match1, match2])
It will get the round like match1.round + 1.

pablonete commented on app/common/utils.js:
```diff
@@ -0,0 +1,7 @@
+export function spliceItemRandomly(array) {
+    return array.splice(getRandomInt(0, array.length), 1)[0];
```
splice means a section, this method instead returns a single item.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+    pairUpPlayers() {
+        var playersCopy = this.players.slice();
+        for (var match of this.matches) {
+            if (match.round == 1) {
```
This if can be easily avoided using another generation approach. First round matches should be a separate list for both the generator and the tournament.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        for (var match of this.matches) {
+            if (match.round == 1) {
+                match.addPlayer(utils.spliceItemRandomly(playersCopy));
+                match.addPlayer(utils.spliceItemRandomly(playersCopy));
```
There was an implicit requirement so ranks are used when creating the bracket: players with rank 1 and 2 must not play until the final game.

josemigallas replied:
Should rank 1 player automatically go into de last round then? Because otherwise that match could happen anywhere after round 1.

pablonete replied:
I don't understand. Idea is that players with rank 1 and  2 don't play before the final. So they will be on the first round, but not on the same match.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,101 @@
+        for (var i=0; i<this.totalMatches(); i++) {
+            this.matches.push(new match.Match());
+        }
+    }
```
The idea after ghost players is that matches where they are involved are also ghost matches, so the other player must promote to the next round and the ghost match removed.
So draft cannot be randomly, because 2 ghost players cannot be paired together. Using ranks and giving ghost players the lowest will do the trick.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+        mTournament.addPlayer(new player.Player("Ironman", 1));
+        mTournament.addPlayer(new player.Player("Wolverine", 1));
+        mTournament.addPlayer(new player.Player("Thor", 1));
+        mTournament.addPlayer(new player.Player("The Thing", 1));
```
Avoid verbosity. Use a private function in the tests like: createTournament({ playersCount: 6})

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+            } else {
+                expect(match.players.length).toEqual(0);
+            }
+        });
```
Randomness must be avoided on unit tests. We don´t want an approved code to fail randomly. And if we get a failing test, it may or may not fail the next time. Random function should be injected in the generator, so tests could inject a non-random function. Thus we can write better asserts, like Spiderman will play against The Thing.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,145 @@
+                expect(match.players[0] instanceof player.Player).toBe(true);
+            }
+        });
+    });
```
Not required, but just to mention, I´m missing the feature of players winning a match and promoting to next round.

josemigallas committed:
Cleans up app.js and adds new file for tournament business logic

josemigallas committed:
Improves readability changing function name and syntax and import name.

josemigallas committed:
Changes class Player for a new function playerValidator and performs player validation through a validationResult object.

josemigallas committed:
fixes module match and deletes class MatchError

josemigallas committed:
fixes typo

josemigallas committed:
adds method addParent in class Match with test

josemigallas committed:
refactors module tournament.js:
- uses Array.find() instead of private function isPlayerEnrolled()
- refactor checkNumberOfPlayers and hasProperNumberOfPlayers for more readability
- pulls class Player out to its own module for better modularity

josemigallas committed:
fixes player and tournament tests

josemigallas committed:
moves private methods out of class body

josemigallas committed:
deletes unnecessary method

josemigallas committed:
refactors player validation for more readability including a new method, with test.

josemigallas committed:
imports class Match directly to improve readability

josemigallas committed:
adds module utils test

josemigallas committed:
moves 2^n-number of players detection to module utils

josemigallas committed:
improves readability and efficiency of method fillUpWithGhosts adding a new method in module utils, with tests

josemigallas committed:
simulates ghost players by simply extending the array without adding any faked player

josemigallas committed:
tournament's player validation considers rank unique for each player

josemigallas committed:
changes and improves the generation algorith of tournament by deleting ghost players usage and adding extended class ChildMatch

josemigallas committed:
refactors match.spec for more readability

josemigallas committed:
refactors tournament.spec for more readability

josemigallas committed:
removes unnecessary methods and test in module utils

josemigallas committed:
refactors module tournament in a neater way

josemigallas committed:
minor fixes: if players ar odd, best player is placed to automatically go into a higher round

josemigallas committed:
fixes error when checking existing player

josemigallas committed:
tournament sorts its players each time a new player is added

josemigallas committed:
lists players in HTML table sorted by rank

josemigallas committed:
separates tournament setup and progress in two modules and prints round one matches after generating

josemigallas committed:
adds getter in class tournament and updates test

josemigallas committed:
implements dynamic tournament bracket

josemigallas committed:
changes match-generation approach by setting child matches instead of parent matches

josemigallas committed:
fixes onClick callbacks according to new approach

josemigallas committed:
improves usability and avoid matches to finish before it has two players

josemigallas committed:
improves bracket's appearance by adding some black space between match panels

josemigallas committed:
fixes error when spliting players into higher and lower ranks

josemigallas committed:
removes unnecessary parameters

josemigallas committed:
prevents oddPlayers to be calculated twice

josemigallas committed:
collapses add players bar and generate tournament button aften generating it

josemigallas committed:
refactors methos name for better understanding

josemigallas committed:
split the bracket in wider columns

josemigallas committed:
announces tournament winner in final round and disables bracket

josemigallas committed:
delegates alerts handling to a new module tournamentAlerts, using a modal to improve appearance

pablonete commented on app/common/utils.js:
```diff
@@ -0,0 +1,7 @@
+export function shiftItemRandomly(array) {
+    return array.splice(getRandomInt(0, array.length), 1)[0];
```
This method is hard to understand and to validate. Try to make it simpler. It has side effects (splice has, while slice has not). If that´s your intention, then shift is not the best name.
I´d expect shiftItemRandomly([1, 2, 3]) to generate [1, 3, 2](3 was shift left).
Unit testing would also help, this is why we call it spec.

josemigallas replied:
Since shift and splice take and remove an item from an array, I thought both names would be illustrative. This methods takes an item randomly from an array, splicing randomly a length 1 array. I guess that name, "takeItemRandomlyFromArray", would be better.

pablonete replied:
"Take from" is not clarifying the "removal" behavior. Use pop like in stacks?

josemigallas replied:
Pop implicitly refers to the last element of a stack. So does shift (in ES6), but at least _explicitly_ means "move from one place to another". Perhaps the simplest solution would be "shiftItemRandomly**_from**_array, to make clear the intention of extracting an element from an array, thus removing it.

pablonete commented on app/common/utils.js:
```diff
@@ -0,0 +1,7 @@
+    return array.splice(getRandomInt(0, array.length), 1)[0];
+}
+
+export function getRandomInt(min, max) {
```
min is included but max is excluded, correct? Then names are dangerous.

josemigallas replied:
Correct, although it is properly used since the non-inclusive upper bound is array's length. Nevertheless it is much clearer changing its name and including the upper bound because the usage is to randomly select an index of an array, and array.length tends to turn out out of bound errors.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,39 @@
+        this.nextMatch = match;
+    }
+
+    setWinner(player) {
```
Is player expected to be a number? Use index or playerIndex instead.
Or even better, let player be a player.

josemigallas replied:
It would be very natural to get a full player as parameter, but it would also mean a loss of readability in case you don't know which players there are. It would be necessary then to:
1. Check if player is in that match, comparing the name of each one, and in case not, throw.
2. Retrieve the player from the match's players array by means of the index and then call the method with it.
Since there are always 2 players in a match to be set as winner and they are already stored in the match, I found it more important here to simplify code and implementation by renaming the variable and using an index 0-1.

pablonete replied:
You still have to validate indexes, so validating the player belongs to the match is not harder.
And indexes are more error prone. If you pass the wrong player to a match, you detect the bug and throw. If you use indexes, you have less consistency.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,39 @@
+
+    setWinner(player) {
+        if (!this.isFinished) {
+            this.isFinished = true;
```
This way you won´t be able to render the winner. Perhaps this should save the player and not just true.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+export class ChildMatch extends Match {
```
Worst name in the world, unless you use ClassThatExtendsFromMatch. Use domain names instead.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,39 @@
+}
+
+export class ChildMatch extends Match {
+    constructor(firstParent, secondParent) {
```
Same, you can use leftMatch and rightMatch, for example, but don´t use parent, it's not a domain word.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,39 @@
+        if (firstParent.round === 1 && firstParent.players.length === 1) {
+            firstParent.setWinner(0);
+        } else if (secondParent.round === 1 && secondParent.players.length === 1) {
+            secondParent.setWinner(0);
```
Why does this logic belong here? Seems out of place.

josemigallas replied:
Totally true, this logic is used in one single moment and does not depend on class ChildMatch but on whether the number of players is odd.

pablonete commented on app/tournament/player.js:
```diff
@@ -0,0 +1,23 @@
+export function validatePlayer(data) {
+    var validationResult = {};
+
+    validationResult.error = validateName(data.name) + validateRank(data.rank);
```
Don't concatenate, use a list. Let the UI to make the decision of how to show data.

pablonete commented on app/tournament/player.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+function validateRank(rank) {
+    return (rank < 1) ? "Rank has to be greater than 0." : "";
```
Is the string "2" valid as a rank?

josemigallas replied:
Well, I think strings containing valid values should be valid as well so I have to include it in the validator. Also, decimal numbers should not be valid ranks.

pablonete replied:
My point is that the logic is assuming a string like a number without further parsing. I'd be explicit on that, perhaps out of this validation logic, to ensure that the rank is always a number, not a string representing a number.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+
+    addPlayer(player) {
+        this.players.push(new Player(player.name, player.rank));
+        sortPlayersByRank(this);
```
Send just the players.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    addPlayer(player) {
+        this.players.push(new Player(player.name, player.rank));
```
Why do you have to create another player with data from the player?

josemigallas replied:
Just in case the parameter was an object containing name and rank instead of an instance of class Player... but yeah, unnecessary after all

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+        this.matches.push(match);
+    }
+
+    getRounds() {
```
I don't know what to expect from this method. A number? A list of rounds? Even reading the implementation, I don't know. Use unambiguous names.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+    });
+}
+
+function generateTournamentBracket(tournament) {
```
What value adds this function?

josemigallas replied:
I made that method in an attempt to keep the class body short and concise. It has no value in terms of implementation but looks more organised to me that way.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+
+        if (validationResult.success && this.existsPlayer(player)) {
+            validationResult.error = "Player already enrolled";
+            validationResult.success = false;
```
Don't reuse objects. Create new instances, they are cheaper than the errors you may introduce this way.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+}
+
+function generateFirstRoundMatches(tournament) {
+    var oddPlayers = tournament.players.length % 2;
```
Why odd? How is 6 players better than 5?

EDIT: I think I got it know after reading further and seeing the spec. Names are not helping here. For instance, oddPlayers can only be 0 or 1, it's a boolean, not a number or a list somebody could think off.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    it('should return false when searching for a non-enrolled player', function() {
+        expect( tournament.existsPlayer(spiderman) ).toBeUndefined();
```
You say false but return undefined. Undefined is false, yeah, but don't want to specify undefined, you want to assert false. Implemention should not pollute this.

josemigallas replied:
Just a typo, existsPlayer does return undefined if player is not found. Actually, it would be more coherent to return true/false instead of undefined/Player

pablonete replied:
No, I'm fine returning undefined/Player. But that's not the spec, that's the implementation. Assert what you expect, not what you do. You should be able to change from false/true to undefined/object without breaking the spec.

josemigallas replied:
The point is I wrote false instead of undefined. _Nevertheless_, a method that asks "exists player?" should return, in my opinion, yes or no (true/false). It also lacks the spec where true is returned if player is found.

pablonete replied:
In JS, undefined is a valid false value, so you should not write a test where you expect `undefined`, you should expect _any false value_, not exactly `false`.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+
+    function setMockedTournament(players) {
+        while (players--) {
+            tournament.addPlayer(spiderman);
```
Is it OK to have the same player multiple times? I think it's easy to fix.

josemigallas replied:
It just does the trick, since is a mocked tournament, but you are right.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+        tournament.addPlayer(spiderman);
+        var validationResult = tournament.validatePlayer(spiderman);
+
+        expect( validationResult.success ).toBe(false);
```
This can be on previous test, so we avoid duplicating code. Same with the next 2 tests.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+        tournament.addPlayer(spiderman);
+        var validationResult = tournament.validatePlayer(batman);
+
+        expect( validationResult.success ).toBe(true);
```
Why spaces around parens? It doesn't seem your standard for other method calls.

josemigallas replied:
Only in specs, I found sometimes hard to identify which is the assert

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+
+        expect( tournament.players[0].rank ).toEqual(5);
+
+        tournament.addPlayer(batman);
```
Use local data. Global players are creating an unnecessary dependency between tests. If I ever change batman rank to target another test, I'll break this one. It's fragile without any good reason.

josemigallas replied:
True, but the reason is precisely because those players are not to be modified, since they represent each one a rank (highest, middle and lowest). Otherwise I would have to instantiate spiderman up to 8 times and batman 4.

pablonete replied:
That's fine for most tests, but not for this specific one that asserts rank values.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+        expect( tournament.matches[0].round ).toEqual(1);
+        expect( tournament.matches[0].nextMatch.round ).toEqual(2);
+        expect( tournament.matches[1].round ).toEqual(1);
+        expect( tournament.matches[1].nextMatch.round ).toEqual(2);
```
I'm not confortable with matches[0] and matches[1]. I don't think that's part of our spec but just an implementation detail. We could change algorithm and matches get sorted differently, and that would break the test without reason. Verify only what needs to be verified.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+        //      match[1]:(spiderman[0] vs ...[1])   ->  spidey automatically goes into next round
+        //
+        //      Round 2:
+        //          match[2]:(spiderman[0] vs ...[1])
```
If you use comments, your code is not good enough describing itself. Work it out.

josemigallas replied:
The thing is I think that spec is even too planned and forced so it cannot help being a little indescribable, from my point of view

pablonete replied:
I don't got you.

pablonete commented on spec/app/tournament/tournament.spec.js:
```diff
@@ -0,0 +1,121 @@
+    it('should have a final match between Spiderman and Batman if The Thing is defeated during first round', function() {
+        tournament.addPlayer(theThing);
+        tournament.addPlayer(spiderman);
+        tournament.addPlayer(batman);
```
You are implicitly testing the "3 players generate 3 matches", I'd like to see it separately.

You should even have different describes, one to verify the bracket and one to play matches.

pablonete commented on app/tournamentBusiness/tournamentAlerts.js:
```diff
@@ -0,0 +1,40 @@
+    )
+}
+
+export default function showCustomAlert(title, msg) {
```
What does this default mean?

josemigallas replied:
That function is the only one which can be used whenever a user wanted, independently of the other specific message error. But it really does not mean anything, I just have not used any other default export in the whole project.

pablonete replied:
If it's not useful, it's distracting. Remove noise from your code.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+function bestPlayerDoesntPlayFirstRound(tournament) {
+    var match = new Match(1);
+    match.addPlayer(tournament.players[0]);
+    tournament.addMatch(match);
```
This approach is quite risky and unexpected. A new match with just one player? Can you just hook up the player to the next round? Write tests for this, you have one but it´s not enough descriptive.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+    while (bestPlayers.length) {
+        var match = new Match(1);
+        match.addPlayer(utils.shiftItemRandomly(bestPlayers));
+        match.addPlayer(utils.shiftItemRandomly(worstPlayers));
```
removeRandomItem? extractRandomItem?

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+            tournament.matches[i],
+            tournament.matches[i+1]
+        ));
+    }
```
You have 2 paths here: if this is only intended for 2 rounds, this for will always be run once, so it can be removed. If you are addressing 3 rounds, then you have to loop nested. Be specific.

josemigallas replied:
If I understood you correctly, _generateRestOfMatch_ should distinguish between a tournament with 2 rounds and another with more, but then there would be code duplication since the generation algorithm is independent from the number of rounds.

pablonete commented on app/tournamentBusiness/tournamentAlerts.js:
```diff
@@ -0,0 +1,40 @@
+import bootstrap from "bootstrap";
```
Try to name this module like a worker, not like a list. Notifier?

pablonete commented on app/tournamentBusiness/tournamentAlerts.js:
```diff
@@ -0,0 +1,40 @@
+    $("#alertTitle").append(title);
+    $("#alertMessage").append(msg);
+
+    $("#alertModal").modal("show");
```
Who hides custom alerts?

josemigallas replied:
Modals hide themselves. Perhaps it would be better _ShowCustomModal_?

pablonete replied:
Don´t miss the Alert word, but yeah, the point is that I need to look at the implementation to understand it.

pablonete commented on app/tournamentBusiness/tournamentEnrollment.js:
```diff
@@ -0,0 +1,70 @@
+    $("#playerListBody").empty();
+    tournament.players.forEach( function(player) {
+        $("#playerListBody").append(
+            `<tr><td>${player.name.replace(/</g, "&lt;").replace(/>/g, "&gt;")}</td><td>${player.rank}</td></tr>`
```
Do you really need to escape? Then use a library to escape every dangerous character.
But try to keep code clean, use an escapedName variable.

pablonete commented on app/tournamentBusiness/tournamentEnrollment.js:
```diff
@@ -0,0 +1,70 @@
+import bootstrap from "bootstrap";
```
Why is this into tournamentBusiness? This isn't business, this is UI.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+function createOnClickCallback(match, playerIndex) {
```
Use a better name. I see how this creates a callback and it´s used on click handler, but I miss the key point on what this actually does: Set Winner.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+    if (match.players[1]) {
+        html += `${match.players[1].name}`;
+        html += match.isFinished ? "" : ` <a id="match${i}player1">win</a>`;
+    }
```
Can you reuse the code for each player? I´d say you can return a div for each player and then concatenate:
`'<div foo>' + divPlayer1 + divPlayer2 + '</div>'`
This makes logic easier to read and verify correctness.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+            space += `<div class="panel panel-default box-bracket-ghost-med"></div>`;
+        }
+        if (round > 2) {
+            space += `<div class="panel panel-default box-bracket-ghost"></div>`.repeat(round - 2);
```
This approach is original but hard to validate. Have you considered an HTML table? No need to revamp, just think on it.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+function appendAllMatchesWithSpacesToHtml() {
+    tournament.matches.forEach( function(match, i, matches) {
+        if (match.round > 1) {
+            appendTransparenSpacesToHtml(match.round, matches[i-1].round);
```
This is very fragile because it's based upon the order of the matches. It will fail if matches are reversed, which at the model level would be legit. Try to understand well what's the public contract and what are implementation details you should not base on.

It's also a bad smell to use forEach and then access previous match. You could use a local state variable, but then I would not use forEach() but for loop, because you are not using a pure method.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+function cleanBracket() {
+    $("#tournamentBracket").empty();
```
Try to inject this element, so your module is more generic. Imagine your consumer says: "render here".

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+function appendTransparenSpacesToHtml(round, lastRound) {
+    $(`#round${round}`).append(
```
These IDs make the code also fragile. I think you could avoid them by keeping the elements when they are generated.

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+
+    html += "</div></div>";
+
+    return html;
```
Most of this code, even being UI, could be easily tested if you abstract usages to JQuery. Basically it's like
`expects(generateMatch(match)).toBe('<div foo>...');`

pablonete commented on app/tournamentBusiness/tournamentProgress.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+function generateMatchHtmlPanel(match, i) {
+    var html = `<div class="panel panel-default box-bracket"><div>`;
```
What's i? Name better.
And can you avoid it?

pablonete commented:
Looks good overall. I see some points you have to improve.
- Abstract out implementation details.
- By better defining the contract of your modules and classes. It's very important for the domain model.
- Use better names. You are descriptive, but some names are not free of ambiguity.
  Let me know if you want another pass.

josemigallas committed:
fixes error: getRandomInt should include both limits min and max

josemigallas committed:
refactors method for better understanding

josemigallas committed:
fixes error: array out of bound due to getRandomInt refactoring

josemigallas committed:
adds utils specs

josemigallas committed:
changes class Match to store its winner since it is a more useful value than isFinished

josemigallas committed:
renames class ChildMatch and arguments to a more domain name

josemigallas committed:
updates Match specs

josemigallas committed:
improves player validation and tests by adding more conditions

josemigallas committed:
validation errors are stored as a list instead of concatenating strings

josemigallas committed:
fixes typo

josemigallas committed:
erases unnecessary logic
rank 1 player has to be set winner manually from round 1 if it is alone

josemigallas committed:
removes players sorting from addPlayer method and does it while generating

josemigallas committed:
removes unnecessary check

josemigallas committed:
renames umbiguous name getRounds

josemigallas committed:
fixes typo

josemigallas committed:
renames variable in a more clarifying way

josemigallas committed:
improves method existsPlayer: it should return true or false

josemigallas committed:
improves robustness of method setMockedTournament

josemigallas committed:
removes code duplication

josemigallas committed:
removes comment

josemigallas committed:
refactors spec in a more clarifying way

josemigallas committed:
removing noise from code

josemigallas committed:
refactors getting a player: object playerData is equivalent to a Player instance

josemigallas committed:
renames module with more proper name

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    setWinner(playerIndex) {
+        if (this.winner === undefined) {
```
Don´t fail silently. If winner is already defined, calling this method is a bug.

pablonete replied:
Btw don´t compare with undefined. Use this.winner as a boolean.

pablonete commented on app/tournament/match.js:
```diff
@@ -0,0 +1,32 @@
+    setWinner(playerIndex) {
+        if (this.winner === undefined) {
+            this.winner = this.players[playerIndex];
+            this.nextMatch.addPlayer(this.players[playerIndex]);
```
If I set the winner in the final match, nextMatch won´t exist.

pablonete commented on app/tournament/player.js:
```diff
@@ -0,0 +1,24 @@
+
+    if (parseFloat(player.rank) === NaN
+        || !Number.isInteger(parseFloat(player.rank))
+        || parseFloat(player.rank) < 1) {
```
parseInt?
Anyway, don´t parse twice.

pablonete commented on app/tournament/tournament.js:
```diff
@@ -0,0 +1,93 @@
+        return undefined !== this.players.find( function(p) {
+            return p.name.toLowerCase() === this.name.toLowerCase()
+            || p.rank === this.rank;
+        }, player);
```
You can use player in the function and avoid this second argument. Thus it will read simpler.

pablonete commented on spec/app/tournament/player.spec.js:
```diff
@@ -0,0 +1,58 @@
+        expect( validationResult1.error.pop() ).toEqual("Rank has to be an integer number greater than 0.");
+        expect( validationResult2.error.pop() ).toEqual("Rank has to be an integer number greater than 0.");
+        expect( validationResult3.error.pop() ).toEqual("Rank has to be an integer number greater than 0.");
+        expect( validationResult4.error.pop() ).toEqual("Rank has to be an integer number greater than 0.");
```
This organization hurts readability. Act and Assert, Act and Assert...

pablonete commented on spec/app/tournament/player.spec.js:
```diff
@@ -0,0 +1,58 @@
+        var validationResult1 = validatePlayer(new Player("Spiderman", "two"));
+        var validationResult2 = validatePlayer(new Player("Spiderman", undefined));
+        var validationResult3 = validatePlayer(new Player("Spiderman", NaN));
+        var validationResult4 = validatePlayer(new Player("Spiderman", null));
```
Numbered variables are error prone. Use meaningful names, or even inline variables.
